### PR TITLE
Adds the _java_runtime attribute to fix toolchain use in Scala.

### PR DESCRIPTION
**Pull request by sethkoehler**
When executing non-locally, the wrong version of Java would be used,
referencing a path that didn't exist. This change adds the
_java_runtime attribute to have to locate the java runtime on the
machine where the Scala code is being executed, and selects the
correct configuration for the host_javabase.

As a minor cleanup, this also removes the now no-longer-necessary
_java attribute.

TESTED=Ran unit tests and also ran a custom Scala unit test through:
previously, the Scala unit test failed with an incorrect java path
error, now it passes.

Change-Id: I9c2829bded04ec2d9d014094bd9e72cd984fe6e6

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by sethkoehler** [comment]
I think the bazel-io comment is automated, but I believe the admins have already okayed this change, but let me know if that's not the case!

**Comment by ittaiz** [comment]
Test this please (ci bot)

**Comment by sethkoehler** [comment]
Hi there!  Sorry, not sure *how* to test this?  Is there a doc that indicates what tests I need to run that I missed?  I tried "bazel test ..." but that gave errors from the beginning :(

**Comment by sethkoehler** [comment]
Note, I should have mentioned, I did run "bazel test test/..." already, if that is what you're referring to (all tests pass)?

**Comment by ittaiz** [comment]
Sorry, that was for the ci bot to run.

**Comment by sethkoehler** [comment]
Ah no worries, I'd misunderstood, thanks!

File: scala/scala.bzl [hunk 1]
```diff
@@ -409,13 +409,33 @@ def write_manifest(ctx):
...
+    return False
+
 def _write_launcher(ctx, rjars, main_class, jvm_flags, args="", wrapper_preamble=""):
     runfiles_root = "${TEST_SRCDIR}/%s" % ctx.workspace_name
     # RUNPATH is defined here:
     # https://github.com/bazelbuild/bazel/blob/0.4.5/src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt#L227
     classpath = ":".join(["${RUNPATH}%s" % (j.short_path) for j in rjars])
     jvm_flags = " ".join([ctx.expand_location(f, ctx.attr.data) for f in jvm_flags])
-    javabin = "%s/%s" % (runfiles_root, ctx.executable._java.short_path)
+    # TODO: Replace the following if/else with just .java_executable_runfiles_path
```
Comment (ittaiz) [inline scala/scala.bzl]: do we know when this will be available? is there a relevant issue or PR we can link to it?

File: test/BUILD [hunk 2]
```diff
@@ -214,6 +214,12 @@ scala_library(
     name = "MixJavaScalaLib",
     srcs = glob(["src/main/scala/scala/test/mix_java_scala/*.scala"]) +
            glob(["src/main/scala/scala/test/mix_java_scala/*.java"]),
+    javac_jvm_flags = [
```
Comment (ittaiz) [inline test/BUILD]: Was this added for demonstration/regression or is this needed?
If this is needed I'd love to understand why.
If this is for the former reason then it's also unclear to me how this protects from regression

File: scala/scala.bzl [hunk 3]
```diff
@@ -730,8 +750,8 @@ def _scala_binary_common(ctx, cjars, rjars, transitive_compile_time_jars, jars2l
 
   # _java_toolchain added manually since _java doesn't currently setup runfiles
```
Comment (ittaiz) [inline scala/scala.bzl]: is this comment incorrect anymore?

**Comment by johnynek** [comment]
@sethkoehler did you use this with remote execution?

I think @ianoc has seen similar bugs, but we didn't track it all the way down.

cc @gkk-stripe @andyscott

File: scala/scala.bzl (hunk 1 above)
Comment (sethkoehler) [reply scala/scala.bzl]: Would the bazel commit be acceptable?  The property is already in, but I felt it a bit premature to depend on it without it being in a bazel release...and even then, I suspect waiting for a while until folks are generally using those releases would be good:

https://github.com/bazelbuild/bazel/commit/f2075d27ca124156fcd7c01242c552175c0cf145

**Comment by sethkoehler** [comment]
@johnynek 

Yep, this indeed gets tests to work with remote execution (I tested that path explicitly).

**Comment by ianoc** [comment]
@sethkoehler what is the error you saw?

File: scala/scala.bzl (hunk 1 above)
Comment (ittaiz) [reply scala/scala.bzl]: sounds good

**Comment by sethkoehler** [comment]
@ianoc 

I got this error when trying to run a Scala unit test:

<PATH>/bazel-out/k8-fastbuild/bin/src/it/scala/root.runfiles/other/src/it/scala/root_wrapper.sh: line 4: <PATH>/bazel-out/k8-fastbuild/bin/src/it/scala/root.runfiles/other/../local_jdk/bin/java: No such file or directory

Basically, the wrong version of Java (the host java) was being used rather than the copy in the toolchain.

**Comment by ianoc** [comment]
Yeah i'm afraid thats actually a bug in the remote execution stuff not in the rules, we shouldn't have absolute JVM paths in the rules i believe.

The bug has been i guess fixed on master, but you need to enable a few experimental options to have good times with the cache :
```
build --spawn_strategy=sandboxed --genrule_strategy=sandboxed --experimental_strict_action_env --experimental_multi_threaded_digest --strategy=Scalac=worker --jobs=30 --worker_max_instances 10 --experimental_remote_spawn_cache --remote_rest_cache=http://mycache
```
are the configs i'm using now. Post 0.9.0 i think this should fix it for you

**Comment by ianoc** [comment]
https://bazel-review.googlesource.com/c/bazel/+/23350
is the CL for handling this. this effects other rules too than the scala ones, but from what i've seen this fix should make all of them work

**Comment by sethkoehler** [comment]
@ianoc 

Ah, so you're saying this change is likely not needed at all once the next version of bazel is released?  If so, I can withdraw this change!

**Comment by ianoc** [comment]
I think the change itself might be a bug to try get around the bazel bug and might cause issues later. Since our outputs would depend on absolute local jvm paths that sounds like it would cause caching issues in future?

Bazel's current release has the fix in it, but its behind an experimental flag, unfortunately bazel seems to keep things behind these flags for a long time. But if you are on >= 0.9.0 and try your repo with the `--experimental_remote_spawn_cache` i think it should clear up for you.

**Comment by sethkoehler** [comment]
Thanks for the heads up about that!  Let me look into this and get back with you...using that flag seems to force me to run the test locally, which isn't what I was aiming for.

**Comment by ianoc** [comment]
I've mostly been using remote caching rather than execution, if the remote execution starts breaking then i suspect thats a new bug. ulf @ google is very responsive about this and good for a fix!

**Comment by sethkoehler** [comment]
Awesome, I'll discuss with bazel folks, thanks!

**Comment by lberki** [comment]
I think this change is the right thing to do; the way the `_java` attribute was handled before in the Scala rules was incorrect, because they just used whatever JDK Bazel was running under (`@bazel_tools//tools/jdk:java` is an alias to `@local_jdk//:java`). This change fixes that and uses the Java runtime in the target configuration (that's what `@bazel_tools//tools/jdk:current_java_runtime` is), which is exactly how it should be.

**Comment by lberki** [comment]
https://bazel-review.googlesource.com/c/bazel/+/23350 is orthogonal to this change; it deals with execution requirements for action, which is unrelated to the JVM paths set in the action. As far as I am concerned, this PR does the right thing: uses the path to the JVM in the target configuration and not the path to the JVM Bazel is running under.

**Comment by ianoc** [comment]
They are orthogonal changes, but the error mentioned/what folks have ran into before has been caused by the runfiles not being unpacked correctly when being restored from a cache. So in previous debugging the path for the jvm itself was a red-herring. 

if the code here is wrong then by all means we should update it.

**Comment by johnynek** [comment]
Well, if @lberki says this is the right way, and tests pass, it's good enough for me.

**Comment by ittaiz** [comment]
I agree with Oscar.
@sethkoehler any chance you can reply on the two pending comments?

File: scala/scala.bzl (hunk 3 above)
Comment (sethkoehler) [reply scala/scala.bzl]: Good catch, this comment is incorrect (as is one above it as well).  I'm actually checking into whether I need _java_toolchain here at all...will update when I have an answer (should be by tomorrow).

File: test/BUILD (hunk 2 above)
Comment (sethkoehler) [reply test/BUILD]: Thanks for catching this as well, this was not needed after all (I'd mistakenly thought it was).  I reverted the file.

File: scala/scala.bzl (hunk 3 above)
Comment (sethkoehler) [reply scala/scala.bzl]: Just to make things clear, I've updated the two comments I intend to update once I've resolved the _toolchain question.

File: scala/scala.bzl (hunk 1 above)
Comment (sethkoehler) [reply scala/scala.bzl]: I added this to the comment, thanks!

**Comment by johnynek** [comment]
👍

**Comment by johnynek** [comment]
test this please

**Comment by sethkoehler** [comment]
Ok, sorry for the delay...after further discussion with lberki and trying out an example Scala test with remote execution, we don't believe the comment (nor the _java_toolchain entry) is necessary, so I've removed them from the two locations where I found them.

**Comment by ittaiz** [comment]
Test this please (ci bot)

**Comment by johnynek** [comment]
seems like the google ci isn't responding...

@ittaiz I'm happy to merge as is. What do you think?

**Comment by ittaiz** [comment]
Ditto

**Pull request merged**